modules/wallet/update.go

```diff
@@ -1,107 +1,243 @@
+		}
+		w.walletTransactions = w.walletTransactions[:lastIndex]
 		return
 	}
+	return
+}
 
-	if scod.Direction == dir {
-		// FundTransaction creates outputs and adds them immediately. They will
-		// also show up from the transaction pool, occasionally causing
-		// repeats. Additionally, outputs that used to exist are not deleted.
-		// If they get re-added, we need to know the age of the output.
-		output, exists := key.outputs[scod.ID]
-		if exists {
-			if !output.spendable {
-				output.spendable = true
-			}
-			return
+// applyWalletTransaction adds a wallet transaction to the wallet transaction
+// history.
+func (w *Wallet) applyWalletTransaction(fundType types.Specifier, uh types.UnlockHash, t types.Transaction, confirmationTime types.Timestamp, oid types.OutputID, value types.Currency) {
+	_, exists := w.keys[uh]
+	if exists {
+		// Sanity check - the output should not exist in the wallet transaction
+		// map, this should be the first time it was created.
+		wtid := modules.CalculateWalletTransactionID(t.ID(), oid)
+		_, exists := w.walletTransactionMap[wtid]
+		if exists && build.DEBUG {
+			panic("a wallet transaction is being added for the second time")
 		}
+		wt := modules.WalletTransaction{
+			TransactionID:         t.ID(),
+			ConfirmationHeight:    w.consensusSetHeight,
+			ConfirmationTimestamp: confirmationTime,
+			Transaction:           t,
 
-		// Add the output. Age is set to 0 because the output has not been
-		// spent yet.
-		ko := &knownOutput{
-			id:     scod.ID,
-			output: scod.SiacoinOutput,
-
-			spendable: true,
-			age:       0,
-		}
-		key.outputs[scod.ID] = ko
-	} else {
-		if build.DEBUG {
-			_, exists := key.outputs[scod.ID]
-			if !exists {
-				panic("trying to delete an output that doesn't exist?")
-			}
+			FundType:       fundType,
+			OutputID:       oid,
+			RelatedAddress: uh,
+			Value:          value,
 		}
-		key.outputs[scod.ID].spendable = false
+		w.walletTransactions = append(w.walletTransactions, wt)
+		w.walletTransactionMap[wtid] = &w.walletTransactions[len(w.walletTransactions)-1]
+		w.historicOutputs[oid] = value
+		return
 	}
+	return
 }
 
+// ProcessConsensusChange parses a consensus change to update the set of
+// confirmed outputs known to the wallet.
 func (w *Wallet) ProcessConsensusChange(cc modules.ConsensusChange) {
-	// TODO: Restruture whole wallet.
-
-	// Adjust the confirmed set of diffs.
-	for _, scod := range cc.SiacoinOutputDiffs {
-		w.applyDiff(scod, modules.DiffApply)
+	// There are two different situations under which a subscribee calls
+	// ProcessConsensusChange. The first is when w.subscribed is set to false
+	// AND the mutex is already locked. The other situation is that subscribed
+	// is set to true and is not going to be changed. Therefore there is no
+	// race condition here. If w.subscribed is set to false, trying to grab the
+	// lock would cause a deadlock.
+	if w.subscribed {
+		lockID := w.mu.Lock()
+		defer w.mu.Unlock(lockID)
```

Comment:
Yeah there really wasn't anything else I could think of. The wallet needed a lock during the call to 'Unlock', but the call to 'Unlock' triggers a consensus subscription, which calls 'ProcessConsensusChange', which also needs a lock.

This wouldn't have been necessary with the old, threaded model of subscriptions.